### ISSUE TITLE
Step 14 Early Stopping - όχι "U" shape για validation loss

### ISSUE THREAD

**Issue by michaelbzms** [slp-ntua/patrec-labs#62]
Κάνουμε training στο step 14 και δοκιμάζοντας early stopping με μεγάλο patience (πχ 100 epochs μετά το καλύτερο σημείο σταματημού) (καθώς έχουμε checkpoints) το validation loss, ενώ κάνει κάποια spikes, δεν φαίνεται να αυξάνεται συστηματικά ποτέ έτσι ώστε να δημιουργήσει το αναμενόμενο "U" shape. Αρχικά, σκεφτήκαμε ότι ίσως φταίει το L2 Regularization ή/και το dropout αλλά και χωρίς αυτά το ίδιο κάνει. Ενδεικτικά, δείτε τα παρακάτω learning curves:

![image](https://user-images.githubusercontent.com/35953365/102993247-77be5180-4525-11eb-9d01-fe69b0670faf.png)

![image](https://user-images.githubusercontent.com/35953365/102993861-a0931680-4526-11eb-9e5d-479ca1b7385d.png)

Υπάρχει κάποια εξήγηση γιατί συμβαίνει αυτό; 

Παρατηρούμε ότι όσο αυξάνουμε το patience (σε epochs) τόσο καλύτερο τελικό αποτέλεσμα παίρνουμε με το καλύτερο μέχρι στιγμής να είναι 99.67% accuracy στο test set (μόνο 1 λάθος). Και όλα αυτά με το μη bidirectional.

**Comment by georgepar** [comment]
Γενικά δεν υπάρχει κάποιος αυστηρός κανόνας για τη μορφή της καμπύλης του loss. Η μορφή του εξαρτάται από το πόσο εύκολο είναι το πρόβλημα, πόσο εκφραστικό είναι το μοντέλο και πόσο επιθετικός είναι ο optimizer (sgd πιο αργό και ομαλό convergence, adam πιο γρήγορο με περισσότερα σκαμπανευάσματα). Στην περίπτωσή σας έχετε ένα εύκολο πρόβλημα και τα δεδομένα δεν είναι θορυβώδη, οπότε βλέπετε αυτή τη συμπεριφορά. 

Παρόλα τις παραμέτρους που έχετε θέσει δεν θα τις επέλεγα στο 90% των πρακτικών εφαρμογών. Στην πρώτη φάση ανάπτυξης ενός μοντέλου ο γενικός στόχος είναι να κάνεις όσο λιγότερο training γίνεται με όσο το δυνατό καλύτερη επίδοση, για να αποφύγεις το overfitting, που είναι και ο κύριος λόγος αποτυχίας ενός overparameterized μοντέλου (πχ ένα νευρωνικό). Σε αυτό το πνεύμα εγώ θα έκοβα την εκπαίδευση γύρω στην εποχή 70 που αρχίζουν τα πρώτα spikes και έχει πρακτικά συγκλίνει το μοντέλο. Για αυτό το λόγο και το patience συνήθως παίζει στο 3-10 όχι στο 100.

**Comment by michaelbzms** [comment]
Ναι, τελικα στο 10 το αφησαμε. Το εβαλα στο 100 για να δω αν θα ανεβει το validation loss. Adam χρησιμοποιούμε. Στο bidirectional ειχε περισσοτερα "σκαμπανευάσματα".
